src/System.IO.Pipelines/tests/PipeReaderWriterFacts.cs

```diff
@@ -304,6 +307,8 @@ void ThrowTestException()
             invalidOperationException = await Assert.ThrowsAsync<InvalidOperationException>(async () => await _pipe.Reader.ReadAsync());
             Assert.Equal("Writer exception", invalidOperationException.Message);
             Assert.Contains("ThrowTestException", invalidOperationException.StackTrace);
+
+            Assert.Single(Regex.Matches(invalidOperationException.StackTrace, "Pipe.GetReadResult"));
```

Comment:
Do you also want a test that if the same source exception is passed into multiple pipes it only returns a single entry? https://github.com/aspnet/KestrelHttpServer/issues/2730#issuecomment-407911151

Reply:
But we can't guarantee that. Different instances of EDI would still mutate the same exception object.

Reply:
```csharp
void Main()
{
    Exception exception;
    try
    {
        throw new Exception();
    }
    catch (Exception ex)
    {
        exception = ex;
    }

    var edi1 = ExceptionDispatchInfo.Capture(exception);
    var edi2 = ExceptionDispatchInfo.Capture(exception);

    try
    {
        edi1.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }
    try
    {
        edi1.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }

    try
    {
        edi2.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }
    try
    {
        edi2.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }
}
```
Doesn't seem to?
```
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 48
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 57
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_pnbqpb.cs:line 67
```

Reply:
```

{
    Exception exception;
    try
    {
        throw new Exception();
    }
    catch (Exception ex)
    {
        exception = ex;
    }

    var edi1 = ExceptionDispatchInfo.Capture(exception);

    try
    {
        edi1.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }

    var edi2 = ExceptionDispatchInfo.Capture(exception);
    try
    {
        edi2.Throw();
    }
    catch (Exception ex)
    {
        Console.WriteLine("***");
        Console.WriteLine(ex.ToString());
    }
}

```

Yes it does:

```
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.RunUserAuthoredQuery() in C:\Users\pakrymet\AppData\Local\Temp\LINQPad5\_clqswnhi\query_ztirrw.cs:line 39
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.RunUserAuthoredQuery() in C:\Users\pakrymet\AppData\Local\Temp\LINQPad5\_clqswnhi\query_ztirrw.cs:line 50
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.RunUserAuthoredQuery() in C:\Users\pakrymet\AppData\Local\Temp\LINQPad5\_clqswnhi\query_ztirrw.cs:line 39
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.RunUserAuthoredQuery() in C:\Users\pakrymet\AppData\Local\Temp\LINQPad5\_clqswnhi\query_ztirrw.cs:line 50
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.RunUserAuthoredQuery() in C:\Users\pakrymet\AppData\Local\Temp\LINQPad5\_clqswnhi\query_ztirrw.cs:line 61


```

Reply:
So `Throw` isn't threadsafe as it uses the same exception object
```csharp
Exception exception;
Exception exception1 = null;
Exception exception2 = null;
try
{
    throw new Exception();
}
catch (Exception ex)
{
    exception = ex;
}

var edi1 = ExceptionDispatchInfo.Capture(exception);

try
{
    edi1.Throw();
}
catch (Exception ex)
{
    Console.WriteLine("***");
    exception1 = ex;
    Console.WriteLine(ex.ToString());
}

var edi2 = ExceptionDispatchInfo.Capture(exception);
try
{
    edi2.Throw();
}
catch (Exception ex)
{
    Console.WriteLine("***");
    exception2 = ex;
    Console.WriteLine(ex.ToString());
}

Console.WriteLine("***");
Console.WriteLine(ReferenceEquals(exception, exception1));
Console.WriteLine(ReferenceEquals(exception, exception2));
```

```
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_tugyst.cs:line 38
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_tugyst.cs:line 49
***
System.Exception: Exception of type 'System.Exception' was thrown.
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_tugyst.cs:line 38
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_tugyst.cs:line 49
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at UserQuery.Main() in C:\Users\thund\AppData\Local\Temp\LINQPad5\_jffesacy\query_tugyst.cs:line 61
***
True
True
```

So it looks like its equivalent to `throw ex`; but it keeps the StackTrace at capture and appends from there; whereas `throw ex` would completely overwrite.